### ISSUE TITLE
.NET 9 SDK introduces hang in `dotnet format` when run on Azure Pipelines in combination with NB.GV

### ISSUE THREAD

**Issue by AArnott** [dotnet/sdk#44951]
### Describe the bug

`dotnet format --verify-no-changes --no-restore` just hangs with no output when run in this configuration:

- Azure Pipelines (local is fine)
- .NET 9 SDK (.NET 8 SDK is fine)
- With the Nerdbank.GitVersioning nuget package referenced

[Working fine with .NET 8 SDK](https://dev.azure.com/andrewarnott/OSS/_build/results?buildId=10813&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=d9b15392-e4ce-5e4c-0c8c-b69645229181)
[Hanging (across 2 attempts) with .NET 9 SDK](https://dev.azure.com/andrewarnott/OSS/_build/results?buildId=10814&view=logs&j=50a69d0a-7972-5459-cdae-135ee6ebe312&t=13df7b5c-76db-5c26-6592-75581a9ed64a)

![Image](https://github.com/user-attachments/assets/063893ef-ffb7-4bec-9c85-f9800c360dc9)

@nietras [reported the same issue with their own repo](https://github.com/dotnet/Nerdbank.GitVersioning/issues/1119), who isolated the repro to requiring both .NET 9 SDK *and* the Nerdbank.GitVersioning package. 

In my case, the hang is on a linux build agent. For @nietras I believe it was a Windows agent. Both of us run this tool only on one agent in the pipeline.
 
### To Reproduce

1. Fork the repo at https://github.com/dotnet/Nerdbank.Streams.git
1. Run it under Azure Pipelines on this commit: 09b404915c9a6d4566023e3bdebbbf8b08eebf42

### Further technical details

@nietras also shared the diagnostic output:

```
D:\agent\_work\_tool\dotnet\dotnet.exe format --verify-no-changes -v diagnostic
  The dotnet runtime version is '9.0.0'.
  The dotnet CLI version is '9.0.100'.
  Using MSBuild.exe located in 'D:\agent\_work\_tool/dotnet\sdk\9.0.100\'.
  Formatting code files in workspace 'D:\agent\_work\210\s\Test.sln'.
  Loading workspace.
    Determining projects to restore...
  Restored D:\agent\_work\210\s\src\XYZ\XYZ.csproj (in 198 ms)
```

<details>
<summary>`dotnet --info` output</summary>

```
 MSBuild version:   17.12.7+5b8665660

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  22.04
 OS Platform: Linux
 RID:         linux-x64
 Base Path:   /opt/hostedtoolcache/dotnet/sdk/9.0.100/

.NET workloads installed:
There are no installed workloads to display.
Configured to use loose manifests when installing new manifests.

Host:
  Version:      9.0.0
  Architecture: x64
  Commit:       9d5a6a9aa4

.NET SDKs installed:
  9.0.100 [/opt/hostedtoolcache/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 9.0.0 [/opt/hostedtoolcache/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 8.0.11 [/opt/hostedtoolcache/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 9.0.0 [/opt/hostedtoolcache/dotnet/shared/Microsoft.NETCore.App]

Other architectures found:
  None

Environment variables:
  DOTNET_ROOT       [/opt/hostedtoolcache/dotnet]

global.json file:
  /home/vsts/work/1/s/global.json

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
Mono JIT compiler version 6.12.0.200 (tarball Tue Jul 11 21:37:50 UTC 2023)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           __thread
	SIGSEGV:       altstack
	Notifications: epoll
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug 
	Interpreter:   yes
	LLVM:          yes(610)
	Suspend:       hybrid
	GC:            sgen (concurrent by default)
```
</details>

**Comment by AArnott** [comment]
Thanks for the suggestion, @kalleolaviniemitalo. [It](https://github.com/dotnet/Nerdbank.Streams/commit/e431e723ee0ee6cfdb309e3b411dd454f3eafb47) [didn't work](https://dev.azure.com/andrewarnott/OSS/_build/results?buildId=10817&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=d9b15392-e4ce-5e4c-0c8c-b69645229181) though.

**Comment by AArnott** [comment]
[This workaround](https://github.com/dotnet/Nerdbank.Streams/pull/800/commits/20e285c4d956424d2686261faa32f8721d5a7acf) has proven effective, but hacky.
TLDR: make NB.GV a conditional dependency, and set an env variable to turn it off around the `dotnet format` command invocation.

**Comment by nietras** [comment]
@AArnott thank you for reporting the issue here. I am indeed seeing the same on Windows on in-house build servers with the following info.
```
D:\agent\_work\_tool\dotnet\dotnet.exe --info
.NET SDK:
 Version:           9.0.100
 Commit:            59db016f11
 Workload version:  9.0.100-manifests.3068a692
 MSBuild version:   17.12.7+5b8665660

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19045
 OS Platform: Windows
 RID:         win-x64
 Base Path:   D:\agent\_work\_tool\dotnet\sdk\9.0.100\

.NET workloads installed:
There are no installed workloads to display.
Configured to use loose manifests when installing new manifests.

Host:
  Version:      9.0.0
  Architecture: x64
  Commit:       9d5a6a9aa4
```
It is very easy to reproduce by creating a small csproj that references `Nerdbank.GitVersioning`
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net9.0</TargetFramework>
    <Nullable>enable</Nullable>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Nerdbank.GitVersioning" Version="3.6.143" 
                      PrivateAssets="all" Condition="!Exists('packages.config')" />
  </ItemGroup>

</Project>
```
with `global.json`:
```json
{
  "sdk": {
    "version": "9.0.100",
    "rollForward": "latestPatch",
    "allowPrerelease": false
  }
}
```
and then defining a pipeline `azure-pipeline.yml` like:
```yml
trigger:
- main

pool:
  name: Default

jobs:
- job: Build_and_Test
  steps:
  - checkout: self
    clean: true
    persistCredentials: true
    fetchDepth: 0
  
  - task: NuGetToolInstaller@1
    displayName: 'Use NuGet 6.4'
    inputs:
      versionSpec: 6.4
  
  - task: NuGetAuthenticate@1
    displayName: 'NuGet Authenticate'
    inputs:
      forceReinstallCredentialProvider: true
  
  - task: UseDotNet@2
    displayName: 'Use .NET - global.json'
    inputs:
      useGlobalJson: true

  - task: DotNetCoreCLI@2
    displayName: 'dotnet --info'
    inputs:
      command: custom
      custom: '--info'

  - task: DotNetCoreCLI@2
    displayName: 'dotnet format verify-no-changes'
    inputs:
      command: custom
      custom: format
      arguments: '--verify-no-changes -v diagnostic'
```
which will then hang with no other output than:
```
The dotnet runtime version is '9.0.0'.
  The dotnet CLI version is '9.0.100'.
  Using MSBuild.exe located in 'D:\agent\_work\_tool/dotnet\sdk\9.0.100\'.
  Formatting code files in workspace 'D:\agent\_work\96\s\Test.sln'.
  Loading workspace.
    Determining projects to restore...
  Restored D:\agent\_work\96\s\src\Some\Some.csproj (in 130 ms).
```

**Comment by christianheld** [comment]
Updating to .NET SDK Version 9.0.200 solved the issue (at least for me).

**Comment by AArnott** [comment]
I can confirm that 9.0.200 seems to have solved it. Thanks for calling that out, @christianheld.

**Comment by AArnott** [comment]
Re-opening the issue, because we have a recurrence of it with .NET SDK 9.0.203.

(note to self: in the validation insertion pipeline of [Pull Request 630650](https://dev.azure.com/devdiv/DevDiv/_git/VisualStudio.Conversations/pullrequest/630650): Merge latest Library.Template)
It failed twice in a row.

**Comment by mslukebo** [comment]
I am also experiencing this issue, but on .NET 8

`dotnet format` ran fine with runtime version is '8.0.15' (SDK version 8.0.408), but is hanging with runtime version is '8.0.16' (SDK version 8.0.409). Hopefully this information is useful in helping determine which change is causing the issue.

**Comment by jasonmalinowski** [comment]
I don't have specific evidence this is the root cause (other than it's a deadlock in the component dotnet format uses), but .NET 10 Preview 6 should carry along https://github.com/dotnet/roslyn/pull/78927 which _might_ fix it. @AArnott I looked at making a quick test to the internal repo that hit this and not sure I can do that simply without larger changes to the repo...
